### ISSUE TITLE
Native modules compiled incorrectly

### ISSUE THREAD

**Issue by lhulet** [apex/apex#468]
- __Operating system__:
OS X 10.11.5
  - __Apex Version__:
0.10.2

I'm using apex to deploy my code. However, modules like bcrypt aren't compiled correctly on my local machine for lambda. I'm getting an "Invalid ELF header" error.

Shouldn't there be a way for apex to automatically support connecting to an Amazon Linux EC2 instance to compile and deploy the code?

Am I missing something, because this seems like a very manual and tedious process to do every time that I want to deploy the code?

**Comment by mooyoul** [comment]
Lambda function will be ran in Amazon Linux Instance. So you should compile/build native modules on Amazon Linux Instance if you are using apex on Mac. You can make your build pipeline by using hooks.

**Comment by tj** [comment]
Yep like @mooyoul mentioned, anything native has to be built on the same platform. For example this is not a problem with Go, because you can cross-compiled on a Mac to the target Linux environment, Node however does not have this features so you're stuck building in CI or in a VM.

**Comment by tj** [comment]
Closing for now, it's a good question but unrelated to apex itself. Might be nice to have in the FAQ though

**Comment by lhulet** [comment]
CI resolved this issue for me. 3.5/4 would recommend.

**Comment by samalba** [comment]
Sorry to comment on a closed issue, I came across this one and had a hard time to come up with a good solution. I wanted something light, without having to setup an entire build system on a Linux box.

I just wanted to share my solution as some people may find it useful. I build my python functions in a Docker container using [Docker for Mac](https://www.docker.com/docker-mac).

Then I simply added this hook to my function.json:

```json
  "hooks":{
    "build": "docker run -v $(pwd):/build --rm -w /build python:3.6 pip3 install -r requirements.txt -t .",
    "clean": "git ls-files --directory --others --exclude-standard | xargs rm -rf ; rm -rf __pycache__ ; rm -f *.so"
  }
```

Then, when you run your `apex deploy`, the python dependencies are built in a temporary container (which runs on Linux). It allows you to still use apex from your Mac.
